AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +604,842 @@ class AzureIoTHub {
-        function receive(done) {
-            _handlers.onEvent <- done;
+        // Opens a connection to Azure IoT Hub.
+        //
+        // Returns:                         Nothing.
+        function connect() {
+            if (_state == S_DISCONNECTED) {
+                _log("Connecting...");
 
-            if (done == null) {
-                _receivers.event.close();
-                // clear out handlers
-                return this;
-            }
+                local devPath = "/" + _connStrParsed.DeviceId;
+                local username = _connStrParsed.HostName + devPath + format("/api-version=%s", AZURE_API_VERSION);
+                local resourcePath = "/devices" + devPath + format("/api-version=%s", AZURE_API_VERSION);
+                local resourceUri = AzureIoTHub.Authorization.encodeUri(_connStrParsed.HostName + resourcePath);
+                local passwDeadTime = AzureIoTHub.Authorization.anHourFromNow();
+                local sas = AzureIoTHub.SharedAccessSignature.create(
+                    resourceUri, null, _connStrParsed.SharedAccessKey, passwDeadTime).toString();
 
-            if( !_isOpen(_sessions.event) ) {
-                _handlers.onEvent("Cannot receive while disconnected.", null);
-            } else {
-                // TODO: test/handle if receiver already open
-                if ( _isOpen(_receivers.event) ) _receivers.event.close();
-                _openEventReceiver();
-            }
-            return this;
-        }
-
-        // PRIVATE FUNCTIONS
-        // ------------------------------------------------------------------------------------
-        // ------------------------------------------------------------------------------------
-
-        // Open Auth/Event Session/Sender/Receivers
-        // ------------------------------------------------------------------------------------
+                local options = {
+                    "username" : username,
+                    "password" : sas
+                };
 
-        function _openAuthSession() {
-            _sessions.auth <- _connection.opensession(_amqpAuthSessionStatusHandler.bindenv(this));
-        }
+                local url = "ssl://" + _connStrParsed.HostName;
 
-        function _openAuthReceiver() {
-            _receivers.auth <- _sessions.auth.openreceiver(AZURE_CLAIM_BASED_SECURITY_PATH, _amqpAuthReceiverStatusHandler.bindenv(this), _authReceiverCB.bindenv(this));
-        }
+                _state =  S_DISCONNECTED | S_CONNECTING;
 
-        function _openAuthSender() {
-            _senders.auth <- _sessions.auth.opensender(AZURE_CLAIM_BASED_SECURITY_PATH, _amqpAuthSenderStatusHandler.bindenv(this));
+                _mqttclient.connect(url, _connStrParsed.DeviceId, options);
+            } else {
+                try {
+                    // We are connected or connecting now
+                    // TODO: Should we call this one and next all callbacks with imp.wakeup?
```

Comment:
no we shouldn't. that's okay to call it synchronously. please remove this TODO

Reply:
Done